CMakeLists.txt

```diff
@@ -235,7 +235,14 @@ file(GLOB SOURCES
 )
 
 add_executable(lightgbm src/main.cpp ${SOURCES})
-add_library(_lightgbm SHARED src/c_api.cpp src/lightgbm_R.cpp ${SOURCES})
+
+# Only build the R part of the library if building for
+# use with the R package
+if(BUILD_FOR_R)
+  add_library(_lightgbm SHARED src/c_api.cpp src/lightgbm_R.cpp ${SOURCES})
+else()
+  add_library(_lightgbm SHARED src/c_api.cpp ${SOURCES})
+endif(BUILD_FOR_R)
```

Comment:
I think the following will a little bit clearer.

```suggestion
add_library(_lightgbm SHARED src/c_api.cpp ${SOURCES})
# Only build the R part of the library if building for
# use with the R package
if(BUILD_FOR_R)
  add_library(_lightgbm SHARED src/lightgbm_R.cpp ${SOURCES})
endif(BUILD_FOR_R)
```

Also, I believe we can link R libraries only for  `_lightgbm` target, because R package doesn't use executable file.
https://github.com/microsoft/LightGBM/blob/0341906c45666ccaeca92ee5fbee17fa22282470/CMakeLists.txt#L310-L313

Reply:
oh I didn't realize you can build it up like that. That's definitely clearer, I agree. And agree with the suggestion to only link into `_lightgbm`. I just tried that locally and it worked well.

Reply:
Seems I was wrong and `add_library` can be used only once for the same target. Instead, `target_sources` should be used, but it's available only in >=3.1 version. So, I think we can create a list of additional sources (refer to `SOURCES`) conditionally before executing `add_library`.

Reply:
ok let me try that

Reply:
tried in https://github.com/microsoft/LightGBM/pull/2911/commits/34f4b0c2ca1970731f7b7df9e1ee28ea6dba3039